// routing-controllers/ActionParameterHandler.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionParameterHandler = void 0;
var class_transformer_1 = require("class-transformer");
var class_validator_1 = require("class-validator");
var BadRequestError_1 = require("./http-error/BadRequestError");
var ParameterParseJsonError_1 = require("./error/ParameterParseJsonError");
var ParamRequiredError_1 = require("./error/ParamRequiredError");
var AuthorizationRequiredError_1 = require("./error/AuthorizationRequiredError");
var CurrentUserCheckerNotDefinedError_1 = require("./error/CurrentUserCheckerNotDefinedError");
var isPromiseLike_1 = require("./util/isPromiseLike");
var ParamNormalizationError_1 = require("./error/ParamNormalizationError");
/**
 * Handles action parameter.
 */
var ActionParameterHandler = /** @class */ (function () {
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function ActionParameterHandler(driver) {
        this.driver = driver;
    }
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Handles action parameter.
     */
    ActionParameterHandler.prototype.handle = function (action, param) {
        var _this = this;
        if (param.type === "request")
            return action.request;
        if (param.type === "response")
            return action.response;
        if (param.type === "context")
            return action.context;
        // get parameter value from request and normalize it
        var value = this.normalizeParamValue(this.driver.getParamFromRequest(action, param), param);
        if (isPromiseLike_1.isPromiseLike(value))
            return value.then(function (value) { return _this.handleValue(value, action, param); });
        return this.handleValue(value, action, param);
    };
    // -------------------------------------------------------------------------
    // Protected Methods
    // -------------------------------------------------------------------------
    /**
     * Handles non-promise value.
     */
    ActionParameterHandler.prototype.handleValue = function (value, action, param) {
        // if transform function is given for this param then apply it
        if (param.transform)
            value = param.transform(action, value);
        // if its current-user decorator then get its value
        if (param.type === "current-user") {
            if (!this.driver.currentUserChecker)
                throw new CurrentUserCheckerNotDefinedError_1.CurrentUserCheckerNotDefinedError();
            value = this.driver.currentUserChecker(action);
        }
        // check cases when parameter is required but its empty and throw errors in this case
        if (param.required) {
            var isValueEmpty = value === null || value === undefined || value === "";
            var isValueEmptyObject = typeof value === "object" && Object.keys(value).length === 0;
            if (param.type === "body" && !param.name && (isValueEmpty || isValueEmptyObject)) { // body has a special check and error message
                return Promise.reject(new ParamRequiredError_1.ParamRequiredError(action, param));
            }
            else if (param.type === "current-user") { // current user has a special check as well
                if (isPromiseLike_1.isPromiseLike(value)) {
                    return value.then(function (currentUser) {
                        if (!currentUser)
                            return Promise.reject(new AuthorizationRequiredError_1.AuthorizationRequiredError(action));
                        return currentUser;
                    });
                }
                else {
                    if (!value)
                        return Promise.reject(new AuthorizationRequiredError_1.AuthorizationRequiredError(action));
                }
            }
            else if (param.name && isValueEmpty) { // regular check for all other parameters // todo: figure out something with param.name usage and multiple things params (query params, upload files etc.)
                return Promise.reject(new ParamRequiredError_1.ParamRequiredError(action, param));
            }
        }
        return value;
    };
    /**
     * Normalizes parameter value.
     */
    ActionParameterHandler.prototype.normalizeParamValue = function (value, param) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (value === null || value === undefined)
                            return [2 /*return*/, value];
                        if (!(typeof value === "object" && ["queries", "headers", "params", "cookies"].some(function (paramType) { return paramType === param.type; }))) return [3 /*break*/, 2];
                        return [4 /*yield*/, Promise.all(Object.keys(value).map(function (key) { return __awaiter(_this, void 0, void 0, function () {
                                var keyValue, ParamType, typeString, _a, _b;
                                return __generator(this, function (_c) {
                                    switch (_c.label) {
                                        case 0:
                                            keyValue = value[key];
                                            if (!(typeof keyValue === "string")) return [3 /*break*/, 2];
                                            ParamType = Reflect.getMetadata("design:type", param.targetType.prototype, key);
                                            if (!ParamType) return [3 /*break*/, 2];
                                            typeString = ParamType.name.toLowerCase();
                                            _a = value;
                                            _b = key;
                                            return [4 /*yield*/, this.normalizeParamValue(keyValue, __assign(__assign({}, param), { name: key, targetType: ParamType, targetName: typeString }))];
                                        case 1:
                                            _a[_b] = _c.sent();
                                            _c.label = 2;
                                        case 2: return [2 /*return*/];
                                    }
                                });
                            }); }))];
                    case 1:
                        _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        if (typeof value === "string") {
                            switch (param.targetName) {
                                case "number":
                                case "string":
                                case "boolean":
                                case "date":
                                    return [2 /*return*/, this.normalizeStringValue(value, param.name, param.targetName)];
                            }
                        }
                        _a.label = 3;
                    case 3:
                        if (!((["number", "string", "boolean"].indexOf(param.targetName) === -1)
                            && (param.parse || param.isTargetObject))) return [3 /*break*/, 5];
                        value = this.parseValue(value, param);
                        value = this.transformValue(value, param);
                        return [4 /*yield*/, this.validateValue(value, param)];
                    case 4:
                        value = _a.sent();
                        _a.label = 5;
                    case 5: return [2 /*return*/, value];
                }
            });
        });
    };
    /**
     * Normalizes string value to number or boolean.
     */
    ActionParameterHandler.prototype.normalizeStringValue = function (value, parameterName, parameterType) {
        switch (parameterType) {
            case "number":
                if (value === "") {
                    throw new ParamNormalizationError_1.InvalidParamError(value, parameterName, parameterType);
                }
                var valueNumber = +value;
                if (valueNumber === NaN) {
                    throw new ParamNormalizationError_1.InvalidParamError(value, parameterName, parameterType);
                }
                return valueNumber;
            case "boolean":
                if (value === "true" || value === "1" || value === "") {
                    return true;
                }
                else if (value === "false" || value === "0") {
                    return false;
                }
                else {
                    throw new ParamNormalizationError_1.InvalidParamError(value, parameterName, parameterType);
                }
            case "date":
                var parsedDate = new Date(value);
                if (Number.isNaN(parsedDate.getTime())) {
                    throw new ParamNormalizationError_1.InvalidParamError(value, parameterName, parameterType);
                }
                return parsedDate;
            case "string":
            default:
                return value;
        }
    };
    /**
     * Parses string value into a JSON object.
     */
    ActionParameterHandler.prototype.parseValue = function (value, paramMetadata) {
        if (typeof value === "string") {
            try {
                return JSON.parse(value);
            }
            catch (error) {
                throw new ParameterParseJsonError_1.ParameterParseJsonError(paramMetadata.name, value);
            }
        }
        return value;
    };
    /**
     * Perform class-transformation if enabled.
     */
    ActionParameterHandler.prototype.transformValue = function (value, paramMetadata) {
        if (this.driver.useClassTransformer &&
            paramMetadata.actionMetadata.options.transformRequest !== false &&
            paramMetadata.targetType &&
            paramMetadata.targetType !== Object &&
            !(value instanceof paramMetadata.targetType)) {
            var options = paramMetadata.classTransform || this.driver.plainToClassTransformOptions;
            value = class_transformer_1.plainToClass(paramMetadata.targetType, value, options);
        }
        return value;
    };
    /**
     * Perform class-validation if enabled.
     */
    ActionParameterHandler.prototype.validateValue = function (value, paramMetadata) {
        var isValidationEnabled = (paramMetadata.validate instanceof Object || paramMetadata.validate === true)
            || (this.driver.enableValidation === true && paramMetadata.validate !== false);
        var shouldValidate = paramMetadata.targetType
            && (paramMetadata.targetType !== Object)
            && (value instanceof paramMetadata.targetType);
        if (isValidationEnabled && shouldValidate) {
            var options = paramMetadata.validate instanceof Object ? paramMetadata.validate : this.driver.validationOptions;
            return class_validator_1.validateOrReject(value, options)
                .then(function () { return value; })
                .catch(function (validationErrors) {
                var error = new BadRequestError_1.BadRequestError("Invalid " + paramMetadata.type + ", check 'errors' property for more info.");
                error.errors = validationErrors;
                error.paramName = paramMetadata.name;
                throw error;
            });
        }
        return value;
    };
    return ActionParameterHandler;
}());
exports.ActionParameterHandler = ActionParameterHandler;
