// routing-controllers/RoutingControllers.js
"use strict";
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoutingControllers = void 0;
var ActionParameterHandler_1 = require("./ActionParameterHandler");
var MetadataBuilder_1 = require("./metadata-builder/MetadataBuilder");
var container_1 = require("./container");
var isPromiseLike_1 = require("./util/isPromiseLike");
var runInSequence_1 = require("./util/runInSequence");
/**
 * Registers controllers and middlewares in the given server framework.
 */
var RoutingControllers = /** @class */ (function () {
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function RoutingControllers(driver, options) {
        this.driver = driver;
        this.options = options;
        /**
         * Global interceptors run on each controller action.
         */
        this.interceptors = [];
        this.parameterHandler = new ActionParameterHandler_1.ActionParameterHandler(driver);
        this.metadataBuilder = new MetadataBuilder_1.MetadataBuilder(options);
    }
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Initializes the things driver needs before routes and middleware registration.
     */
    RoutingControllers.prototype.initialize = function () {
        this.driver.initialize();
        return this;
    };
    /**
     * Registers all given interceptors.
     */
    RoutingControllers.prototype.registerInterceptors = function (classes) {
        var _a;
        var interceptors = this.metadataBuilder
            .buildInterceptorMetadata(classes)
            .sort(function (middleware1, middleware2) { return middleware1.priority - middleware2.priority; })
            .reverse();
        (_a = this.interceptors).push.apply(_a, interceptors);
        return this;
    };
    /**
     * Registers all given controllers and actions from those controllers.
     */
    RoutingControllers.prototype.registerControllers = function (classes) {
        var _this = this;
        var controllers = this.metadataBuilder.buildControllerMetadata(classes);
        controllers.forEach(function (controller) {
            controller.actions.forEach(function (actionMetadata) {
                var interceptorFns = _this.prepareInterceptors(__spreadArrays(_this.interceptors, actionMetadata.controllerMetadata.interceptors, actionMetadata.interceptors));
                _this.driver.registerAction(actionMetadata, function (action) {
                    return _this.executeAction(actionMetadata, action, interceptorFns);
                });
            });
        });
        this.driver.registerRoutes();
        return this;
    };
    /**
     * Registers post-execution middlewares in the driver.
     */
    RoutingControllers.prototype.registerMiddlewares = function (type, classes) {
        var _this = this;
        this.metadataBuilder
            .buildMiddlewareMetadata(classes)
            .filter(function (middleware) { return middleware.global && middleware.type === type; })
            .sort(function (middleware1, middleware2) { return middleware2.priority - middleware1.priority; })
            .forEach(function (middleware) { return _this.driver.registerMiddleware(middleware); });
        return this;
    };
    // -------------------------------------------------------------------------
    // Protected Methods
    // -------------------------------------------------------------------------
    /**
     * Executes given controller action.
     */
    RoutingControllers.prototype.executeAction = function (actionMetadata, action, interceptorFns) {
        var _this = this;
        // compute all parameters
        var paramsPromises = actionMetadata.params
            .sort(function (param1, param2) { return param1.index - param2.index; })
            .map(function (param) { return _this.parameterHandler.handle(action, param); });
        // after all parameters are computed
        return Promise.all(paramsPromises).then(function (params) {
            // execute action and handle result
            var allParams = actionMetadata.appendParams ? actionMetadata.appendParams(action).concat(params) : params;
            var result = actionMetadata.methodOverride ? actionMetadata.methodOverride(actionMetadata, action, allParams) : actionMetadata.callMethod(allParams, action);
            return _this.handleCallMethodResult(result, actionMetadata, action, interceptorFns);
        }).catch(function (error) {
            // otherwise simply handle error without action execution
            return _this.driver.handleError(error, actionMetadata, action);
        });
    };
    /**
     * Handles result of the action method execution.
     */
    RoutingControllers.prototype.handleCallMethodResult = function (result, action, options, interceptorFns) {
        var _this = this;
        if (isPromiseLike_1.isPromiseLike(result)) {
            return result
                .then(function (data) {
                return _this.handleCallMethodResult(data, action, options, interceptorFns);
            })
                .catch(function (error) {
                return _this.driver.handleError(error, action, options);
            });
        }
        else {
            if (interceptorFns) {
                var awaitPromise = runInSequence_1.runInSequence(interceptorFns, function (interceptorFn) {
                    var interceptedResult = interceptorFn(options, result);
                    if (isPromiseLike_1.isPromiseLike(interceptedResult)) {
                        return interceptedResult.then(function (resultFromPromise) {
                            result = resultFromPromise;
                        });
                    }
                    else {
                        result = interceptedResult;
                        return Promise.resolve();
                    }
                });
                return awaitPromise
                    .then(function () { return _this.driver.handleSuccess(result, action, options); })
                    .catch(function (error) { return _this.driver.handleError(error, action, options); });
            }
            else {
                return this.driver.handleSuccess(result, action, options);
            }
        }
    };
    /**
     * Creates interceptors from the given "use interceptors".
     */
    RoutingControllers.prototype.prepareInterceptors = function (uses) {
        return uses.map(function (use) {
            if (use.interceptor.prototype && use.interceptor.prototype.intercept) { // if this is function instance of InterceptorInterface
                return function (action, result) {
                    return container_1.getFromContainer(use.interceptor, action).intercept(action, result);
                };
            }
            return use.interceptor;
        });
    };
    return RoutingControllers;
}());
exports.RoutingControllers = RoutingControllers;
