// routing-controllers/driver/express/ExpressDriver.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpressDriver = void 0;
var ActionMetadata_1 = require("../../metadata/ActionMetadata");
var BaseDriver_1 = require("../BaseDriver");
var AccessDeniedError_1 = require("../../error/AccessDeniedError");
var AuthorizationCheckerNotDefinedError_1 = require("../../error/AuthorizationCheckerNotDefinedError");
var isPromiseLike_1 = require("../../util/isPromiseLike");
var container_1 = require("../../container");
var AuthorizationRequiredError_1 = require("../../error/AuthorizationRequiredError");
var index_1 = require("../../index");
var cookie = require("cookie");
var templateUrl = require("template-url");
/**
 * Integration with express framework.
 */
var ExpressDriver = /** @class */ (function (_super) {
    __extends(ExpressDriver, _super);
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function ExpressDriver(express) {
        var _this = _super.call(this) || this;
        _this.express = express;
        _this.loadExpress();
        _this.app = _this.express;
        return _this;
    }
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Initializes the things driver needs before routes and middlewares registration.
     */
    ExpressDriver.prototype.initialize = function () {
        if (this.cors) {
            var cors = require("cors");
            if (this.cors === true) {
                this.express.use(cors());
            }
            else {
                this.express.use(cors(this.cors));
            }
        }
    };
    /**
     * Registers middleware that run before controller actions.
     */
    ExpressDriver.prototype.registerMiddleware = function (middleware) {
        var _this = this;
        var middlewareWrapper;
        // if its an error handler then register it with proper signature in express
        if (middleware.instance.error) {
            middlewareWrapper = function (error, request, response, next) {
                middleware.instance.error(error, request, response, next);
            };
        }
        // if its a regular middleware then register it as express middleware
        else if (middleware.instance.use) {
            middlewareWrapper = function (request, response, next) {
                try {
                    var useResult = middleware.instance.use(request, response, next);
                    if (isPromiseLike_1.isPromiseLike(useResult)) {
                        useResult.catch(function (error) {
                            _this.handleError(error, undefined, { request: request, response: response, next: next });
                            return error;
                        });
                    }
                }
                catch (error) {
                    _this.handleError(error, undefined, { request: request, response: response, next: next });
                }
            };
        }
        if (middlewareWrapper) {
            // Name the function for better debugging
            Object.defineProperty(middlewareWrapper, "name", {
                value: middleware.instance.constructor.name,
                writable: true
            });
            this.express.use(middlewareWrapper);
        }
    };
    /**
     * Registers action in the driver.
     */
    ExpressDriver.prototype.registerAction = function (actionMetadata, executeCallback) {
        var _a;
        var _this = this;
        // middlewares required for this action
        var defaultMiddlewares = [];
        if (actionMetadata.isBodyUsed) {
            if (actionMetadata.isJsonTyped) {
                defaultMiddlewares.push(this.loadBodyParser().json(actionMetadata.bodyExtraOptions));
            }
            else {
                defaultMiddlewares.push(this.loadBodyParser().text(actionMetadata.bodyExtraOptions));
            }
        }
        if (actionMetadata.isAuthorizedUsed) {
            defaultMiddlewares.push(function (request, response, next) {
                if (!_this.authorizationChecker)
                    throw new AuthorizationCheckerNotDefinedError_1.AuthorizationCheckerNotDefinedError();
                var action = { request: request, response: response, next: next };
                try {
                    var checkResult = _this.authorizationChecker(action, actionMetadata.authorizedRoles);
                    var handleError_1 = function (result) {
                        if (!result) {
                            var error = actionMetadata.authorizedRoles.length === 0 ? new AuthorizationRequiredError_1.AuthorizationRequiredError(action) : new AccessDeniedError_1.AccessDeniedError(action);
                            _this.handleError(error, actionMetadata, action);
                        }
                        else {
                            next();
                        }
                    };
                    if (isPromiseLike_1.isPromiseLike(checkResult)) {
                        checkResult
                            .then(function (result) { return handleError_1(result); })
                            .catch(function (error) { return _this.handleError(error, actionMetadata, action); });
                    }
                    else {
                        handleError_1(checkResult);
                    }
                }
                catch (error) {
                    _this.handleError(error, actionMetadata, action);
                }
            });
        }
        if (actionMetadata.isFileUsed || actionMetadata.isFilesUsed) {
            var multer_1 = this.loadMulter();
            actionMetadata.params
                .filter(function (param) { return param.type === "file"; })
                .forEach(function (param) {
                defaultMiddlewares.push(multer_1(param.extraOptions).single(param.name));
            });
            actionMetadata.params
                .filter(function (param) { return param.type === "files"; })
                .forEach(function (param) {
                defaultMiddlewares.push(multer_1(param.extraOptions).array(param.name));
            });
        }
        // user used middlewares
        var uses = __spreadArrays(actionMetadata.controllerMetadata.uses, actionMetadata.uses);
        var beforeMiddlewares = this.prepareMiddlewares(uses.filter(function (use) { return !use.afterAction; }));
        var afterMiddlewares = this.prepareMiddlewares(uses.filter(function (use) { return use.afterAction; }));
        // prepare route and route handler function
        var route = ActionMetadata_1.ActionMetadata.appendBaseRoute(this.routePrefix, actionMetadata.fullRoute);
        var routeHandler = function routeHandler(request, response, next) {
            return executeCallback({ request: request, response: response, next: next });
        };
        // This ensures that a request is only processed once to prevent unhandled rejections saying
        // "Can't set headers after they are sent"
        // Some examples of reasons a request may cause multiple route calls:
        // * Express calls the "get" route automatically when we call the "head" route:
        //   Reference: https://expressjs.com/en/4x/api.html#router.METHOD
        //   This causes a double execution on our side.
        // * Multiple routes match the request (e.g. GET /users/me matches both @All(/users/me) and @Get(/users/:id)).
        // The following middleware only starts an action processing if the request has not been processed before.
        var routeGuard = function routeGuard(request, response, next) {
            if (!request.routingControllersStarted) {
                request.routingControllersStarted = true;
                return next();
            }
        };
        // finally register action in express
        (_a = this.express)[actionMetadata.type.toLowerCase()].apply(_a, __spreadArrays([
            route,
            routeGuard
        ], beforeMiddlewares, defaultMiddlewares, [
            routeHandler
        ], afterMiddlewares));
    };
    /**
     * Registers all routes in the framework.
     */
    ExpressDriver.prototype.registerRoutes = function () {
    };
    /**
     * Gets param from the request.
     */
    ExpressDriver.prototype.getParamFromRequest = function (action, param) {
        var request = action.request;
        switch (param.type) {
            case "body":
                return request.body;
            case "body-param":
                return request.body[param.name];
            case "param":
                return request.params[param.name];
            case "params":
                return request.params;
            case "session-param":
                return request.session[param.name];
            case "session":
                return request.session;
            case "state":
                throw new Error("@State decorators are not supported by express driver.");
            case "query":
                return request.query[param.name];
            case "queries":
                return request.query;
            case "header":
                return request.headers[param.name.toLowerCase()];
            case "headers":
                return request.headers;
            case "file":
                return request.file;
            case "files":
                return request.files;
            case "cookie":
                if (!request.headers.cookie)
                    return;
                var cookies = cookie.parse(request.headers.cookie);
                return cookies[param.name];
            case "cookies":
                if (!request.headers.cookie)
                    return {};
                return cookie.parse(request.headers.cookie);
        }
    };
    /**
     * Handles result of successfully executed controller action.
     */
    ExpressDriver.prototype.handleSuccess = function (result, action, options) {
        // if the action returned the response object itself, short-circuits
        if (result && result === options.response) {
            options.next();
            return;
        }
        // transform result if needed
        result = this.transformResult(result, action, options);
        // set http status code
        if (result === undefined && action.undefinedResultCode) {
            if (action.undefinedResultCode instanceof Function) {
                throw new action.undefinedResultCode(options);
            }
            options.response.status(action.undefinedResultCode);
        }
        else if (result === null) {
            if (action.nullResultCode) {
                if (action.nullResultCode instanceof Function) {
                    throw new action.nullResultCode(options);
                }
                options.response.status(action.nullResultCode);
            }
            else {
                options.response.status(204);
            }
        }
        else if (action.successHttpCode) {
            options.response.status(action.successHttpCode);
        }
        // apply http headers
        Object.keys(action.headers).forEach(function (name) {
            options.response.header(name, action.headers[name]);
        });
        if (action.redirect) { // if redirect is set then do it
            if (typeof result === "string") {
                options.response.redirect(result);
            }
            else if (result instanceof Object) {
                options.response.redirect(templateUrl(action.redirect, result));
            }
            else {
                options.response.redirect(action.redirect);
            }
            options.next();
        }
        else if (action.renderedTemplate) { // if template is set then render it
            var renderOptions = result && result instanceof Object ? result : {};
            options.response.render(action.renderedTemplate, renderOptions, function (err, html) {
                if (err && action.isJsonTyped) {
                    return options.next(err);
                }
                else if (err && !action.isJsonTyped) {
                    return options.next(err);
                }
                else if (html) {
                    options.response.send(html);
                }
                options.next();
            });
        }
        else if (result === undefined) { // throw NotFoundError on undefined response
            if (action.undefinedResultCode) {
                if (action.isJsonTyped) {
                    options.response.json();
                }
                else {
                    options.response.send();
                }
                options.next();
            }
            else {
                throw new index_1.NotFoundError();
            }
        }
        else if (result === null) { // send null response
            if (action.isJsonTyped) {
                options.response.json(null);
            }
            else {
                options.response.send(null);
            }
            options.next();
        }
        else if (result instanceof Buffer) { // check if it's binary data (Buffer)
            options.response.end(result, "binary");
        }
        else if (result instanceof Uint8Array) { // check if it's binary data (typed array)
            options.response.end(Buffer.from(result), "binary");
        }
        else if (result.pipe instanceof Function) {
            result.pipe(options.response);
        }
        else { // send regular result
            if (action.isJsonTyped) {
                options.response.json(result);
            }
            else {
                options.response.send(result);
            }
            options.next();
        }
    };
    /**
     * Handles result of failed executed controller action.
     */
    ExpressDriver.prototype.handleError = function (error, action, options) {
        if (this.isDefaultErrorHandlingEnabled) {
            var response_1 = options.response;
            // set http code
            // note that we can't use error instanceof HttpError properly anymore because of new typescript emit process
            if (error.httpCode) {
                response_1.status(error.httpCode);
            }
            else {
                response_1.status(500);
            }
            // apply http headers
            if (action) {
                Object.keys(action.headers).forEach(function (name) {
                    response_1.header(name, action.headers[name]);
                });
            }
            // send error content
            if (action && action.isJsonTyped) {
                response_1.json(this.processJsonError(error));
            }
            else {
                response_1.send(this.processTextError(error)); // todo: no need to do it because express by default does it
            }
        }
        options.next(error);
    };
    // -------------------------------------------------------------------------
    // Protected Methods
    // -------------------------------------------------------------------------
    /**
     * Creates middlewares from the given "use"-s.
     */
    ExpressDriver.prototype.prepareMiddlewares = function (uses) {
        var _this = this;
        var middlewareFunctions = [];
        uses.forEach(function (use) {
            if (use.middleware.prototype && use.middleware.prototype.use) { // if this is function instance of MiddlewareInterface
                middlewareFunctions.push(function (request, response, next) {
                    try {
                        var useResult = container_1.getFromContainer(use.middleware).use(request, response, next);
                        if (isPromiseLike_1.isPromiseLike(useResult)) {
                            useResult.catch(function (error) {
                                _this.handleError(error, undefined, { request: request, response: response, next: next });
                                return error;
                            });
                        }
                        return useResult;
                    }
                    catch (error) {
                        _this.handleError(error, undefined, { request: request, response: response, next: next });
                    }
                });
            }
            else if (use.middleware.prototype && use.middleware.prototype.error) { // if this is function instance of ErrorMiddlewareInterface
                middlewareFunctions.push(function (error, request, response, next) {
                    return container_1.getFromContainer(use.middleware).error(error, request, response, next);
                });
            }
            else {
                middlewareFunctions.push(use.middleware);
            }
        });
        return middlewareFunctions;
    };
    /**
     * Dynamically loads express module.
     */
    ExpressDriver.prototype.loadExpress = function () {
        if (require) {
            if (!this.express) {
                try {
                    this.express = require("express")();
                }
                catch (e) {
                    throw new Error("express package was not found installed. Try to install it: npm install express --save");
                }
            }
        }
        else {
            throw new Error("Cannot load express. Try to install all required dependencies.");
        }
    };
    /**
     * Dynamically loads body-parser module.
     */
    ExpressDriver.prototype.loadBodyParser = function () {
        try {
            return require("body-parser");
        }
        catch (e) {
            throw new Error("body-parser package was not found installed. Try to install it: npm install body-parser --save");
        }
    };
    /**
     * Dynamically loads multer module.
     */
    ExpressDriver.prototype.loadMulter = function () {
        try {
            return require("multer");
        }
        catch (e) {
            throw new Error("multer package was not found installed. Try to install it: npm install multer --save");
        }
    };
    return ExpressDriver;
}(BaseDriver_1.BaseDriver));
exports.ExpressDriver = ExpressDriver;
