// routing-controllers/metadata/ActionMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionMetadata = void 0;
/**
 * Action metadata.
 */
var ActionMetadata = /** @class */ (function () {
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function ActionMetadata(controllerMetadata, args, globalOptions) {
        this.globalOptions = globalOptions;
        this.controllerMetadata = controllerMetadata;
        this.route = args.route;
        this.target = args.target;
        this.method = args.method;
        this.options = args.options;
        this.type = args.type;
        this.appendParams = args.appendParams;
        this.methodOverride = args.methodOverride;
    }
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Builds everything action metadata needs.
     * Action metadata can be used only after its build.
     */
    ActionMetadata.prototype.build = function (responseHandlers) {
        var classTransformerResponseHandler = responseHandlers.find(function (handler) { return handler.type === "response-class-transform-options"; });
        var undefinedResultHandler = responseHandlers.find(function (handler) { return handler.type === "on-undefined"; });
        var nullResultHandler = responseHandlers.find(function (handler) { return handler.type === "on-null"; });
        var successCodeHandler = responseHandlers.find(function (handler) { return handler.type === "success-code"; });
        var redirectHandler = responseHandlers.find(function (handler) { return handler.type === "redirect"; });
        var renderedTemplateHandler = responseHandlers.find(function (handler) { return handler.type === "rendered-template"; });
        var authorizedHandler = responseHandlers.find(function (handler) { return handler.type === "authorized"; });
        var contentTypeHandler = responseHandlers.find(function (handler) { return handler.type === "content-type"; });
        var bodyParam = this.params.find(function (param) { return param.type === "body"; });
        if (classTransformerResponseHandler)
            this.responseClassTransformOptions = classTransformerResponseHandler.value;
        this.undefinedResultCode = undefinedResultHandler
            ? undefinedResultHandler.value
            : this.globalOptions.defaults && this.globalOptions.defaults.undefinedResultCode;
        this.nullResultCode = nullResultHandler
            ? nullResultHandler.value
            : this.globalOptions.defaults && this.globalOptions.defaults.nullResultCode;
        if (successCodeHandler)
            this.successHttpCode = successCodeHandler.value;
        if (redirectHandler)
            this.redirect = redirectHandler.value;
        if (renderedTemplateHandler)
            this.renderedTemplate = renderedTemplateHandler.value;
        this.bodyExtraOptions = bodyParam ? bodyParam.extraOptions : undefined;
        this.isBodyUsed = !!this.params.find(function (param) { return param.type === "body" || param.type === "body-param"; });
        this.isFilesUsed = !!this.params.find(function (param) { return param.type === "files"; });
        this.isFileUsed = !!this.params.find(function (param) { return param.type === "file"; });
        this.isJsonTyped = (contentTypeHandler !== undefined
            ? /json/.test(contentTypeHandler.value)
            : this.controllerMetadata.type === "json");
        this.fullRoute = this.buildFullRoute();
        this.headers = this.buildHeaders(responseHandlers);
        this.isAuthorizedUsed = this.controllerMetadata.isAuthorizedUsed || !!authorizedHandler;
        this.authorizedRoles = (this.controllerMetadata.authorizedRoles || []).concat((authorizedHandler && authorizedHandler.value) || []);
    };
    // -------------------------------------------------------------------------
    // Private Methods
    // -------------------------------------------------------------------------
    /**
     * Builds full action route.
     */
    ActionMetadata.prototype.buildFullRoute = function () {
        if (this.route instanceof RegExp) {
            if (this.controllerMetadata.route) {
                return ActionMetadata.appendBaseRoute(this.controllerMetadata.route, this.route);
            }
            return this.route;
        }
        var path = "";
        if (this.controllerMetadata.route)
            path += this.controllerMetadata.route;
        if (this.route && typeof this.route === "string")
            path += this.route;
        return path;
    };
    /**
     * Builds action response headers.
     */
    ActionMetadata.prototype.buildHeaders = function (responseHandlers) {
        var contentTypeHandler = responseHandlers.find(function (handler) { return handler.type === "content-type"; });
        var locationHandler = responseHandlers.find(function (handler) { return handler.type === "location"; });
        var headers = {};
        if (locationHandler)
            headers["Location"] = locationHandler.value;
        if (contentTypeHandler)
            headers["Content-type"] = contentTypeHandler.value;
        var headerHandlers = responseHandlers.filter(function (handler) { return handler.type === "header"; });
        if (headerHandlers)
            headerHandlers.map(function (handler) { return headers[handler.value] = handler.secondaryValue; });
        return headers;
    };
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Calls action method.
     * Action method is an action defined in a user controller.
     */
    ActionMetadata.prototype.callMethod = function (params, action) {
        var controllerInstance = this.controllerMetadata.getInstance(action);
        return controllerInstance[this.method].apply(controllerInstance, params);
    };
    // -------------------------------------------------------------------------
    // Static Methods
    // -------------------------------------------------------------------------
    /**
     * Appends base route to a given regexp route.
     */
    ActionMetadata.appendBaseRoute = function (baseRoute, route) {
        var prefix = "" + (baseRoute.length > 0 && baseRoute.indexOf("/") < 0 ? "/" : "") + baseRoute;
        if (typeof route === "string")
            return "" + prefix + route;
        if (!baseRoute || baseRoute === "")
            return route;
        var fullPath = "^" + prefix + route.toString().substr(1) + "?$";
        return new RegExp(fullPath, route.flags);
    };
    return ActionMetadata;
}());
exports.ActionMetadata = ActionMetadata;
