// routing-controllers/driver/BaseDriver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseDriver = void 0;
var class_transformer_1 = require("class-transformer");
var HttpError_1 = require("../http-error/HttpError");
/**
 * Base driver functionality for all other drivers.
 * Abstract layer to organize controllers integration with different http server implementations.
 */
var BaseDriver = /** @class */ (function () {
    function BaseDriver() {
        // -------------------------------------------------------------------------
        // Public Properties
        // -------------------------------------------------------------------------
        /**
         * Global application prefix.
         */
        this.routePrefix = "";
    }
    // -------------------------------------------------------------------------
    // Protected Methods
    // -------------------------------------------------------------------------
    BaseDriver.prototype.transformResult = function (result, action, options) {
        // check if we need to transform result
        var shouldTransform = this.useClassTransformer // transform only if class-transformer is enabled
            && action.options.transformResponse !== false // don't transform if action response transform is disabled
            && result instanceof Object // don't transform primitive types (string/number/boolean)
            && !(result instanceof Uint8Array // don't transform binary data
                ||
                    result.pipe instanceof Function // don't transform streams
            );
        // transform result if needed
        if (shouldTransform) {
            var options_1 = action.responseClassTransformOptions || this.classToPlainTransformOptions;
            result = class_transformer_1.classToPlain(result, options_1);
        }
        return result;
    };
    BaseDriver.prototype.processJsonError = function (error) {
        var _this = this;
        if (!this.isDefaultErrorHandlingEnabled)
            return error;
        if (typeof error.toJSON === "function")
            return error.toJSON();
        var processedError = {};
        if (error instanceof Error) {
            var name_1 = error.name && error.name !== "Error" ? error.name : error.constructor.name;
            processedError.name = name_1;
            if (error.message)
                processedError.message = error.message;
            if (error.stack && this.developmentMode)
                processedError.stack = error.stack;
            Object.keys(error)
                .filter(function (key) { return key !== "stack" && key !== "name" && key !== "message" && (!(error instanceof HttpError_1.HttpError) || key !== "httpCode"); })
                .forEach(function (key) { return processedError[key] = error[key]; });
            if (this.errorOverridingMap)
                Object.keys(this.errorOverridingMap)
                    .filter(function (key) { return name_1 === key; })
                    .forEach(function (key) { return processedError = _this.merge(processedError, _this.errorOverridingMap[key]); });
            return Object.keys(processedError).length > 0 ? processedError : undefined;
        }
        return error;
    };
    BaseDriver.prototype.processTextError = function (error) {
        if (!this.isDefaultErrorHandlingEnabled)
            return error;
        if (error instanceof Error) {
            if (this.developmentMode && error.stack) {
                return error.stack;
            }
            else if (error.message) {
                return error.message;
            }
        }
        return error;
    };
    BaseDriver.prototype.merge = function (obj1, obj2) {
        var result = {};
        for (var i in obj1) {
            if ((i in obj2) && (typeof obj1[i] === "object") && (i !== null)) {
                result[i] = this.merge(obj1[i], obj2[i]);
            }
            else {
                result[i] = obj1[i];
            }
        }
        for (var i in obj2) {
            result[i] = obj2[i];
        }
        return result;
    };
    return BaseDriver;
}());
exports.BaseDriver = BaseDriver;
