// routing-controllers/metadata-builder/MetadataArgsStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataArgsStorage = void 0;
/**
 * Storage all metadatas read from decorators.
 */
var MetadataArgsStorage = /** @class */ (function () {
    function MetadataArgsStorage() {
        // -------------------------------------------------------------------------
        // Properties
        // -------------------------------------------------------------------------
        /**
         * Registered controller metadata args.
         */
        this.controllers = [];
        /**
         * Registered middleware metadata args.
         */
        this.middlewares = [];
        /**
         * Registered interceptor metadata args.
         */
        this.interceptors = [];
        /**
         * Registered "use middleware" metadata args.
         */
        this.uses = [];
        /**
         * Registered "use interceptor" metadata args.
         */
        this.useInterceptors = [];
        /**
         * Registered action metadata args.
         */
        this.actions = [];
        /**
         * Registered param metadata args.
         */
        this.params = [];
        /**
         * Registered response handler metadata args.
         */
        this.responseHandlers = [];
    }
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Filters registered middlewares by a given classes.
     */
    MetadataArgsStorage.prototype.filterMiddlewareMetadatasForClasses = function (classes) {
        var _this = this;
        return classes
            .map(function (cls) { return _this.middlewares.find(function (mid) { return mid.target === cls; }); })
            .filter(function (midd) { return midd !== undefined; }); // this might be not needed if all classes where decorated with `@Middleware`
    };
    /**
     * Filters registered interceptors by a given classes.
     */
    MetadataArgsStorage.prototype.filterInterceptorMetadatasForClasses = function (classes) {
        return this.interceptors.filter(function (ctrl) {
            return classes.filter(function (cls) { return ctrl.target === cls; }).length > 0;
        });
    };
    /**
     * Filters registered controllers by a given classes.
     */
    MetadataArgsStorage.prototype.filterControllerMetadatasForClasses = function (classes) {
        return this.controllers.filter(function (ctrl) {
            return classes.filter(function (cls) { return ctrl.target === cls; }).length > 0;
        });
    };
    /**
     * Filters registered actions by a given classes.
     */
    MetadataArgsStorage.prototype.filterActionsWithTarget = function (target) {
        return this.actions.filter(function (action) { return action.target === target; });
    };
    /**
     * Filters registered "use middlewares" by a given target class and method name.
     */
    MetadataArgsStorage.prototype.filterUsesWithTargetAndMethod = function (target, methodName) {
        return this.uses.filter(function (use) {
            return use.target === target && use.method === methodName;
        });
    };
    /**
     * Filters registered "use interceptors" by a given target class and method name.
     */
    MetadataArgsStorage.prototype.filterInterceptorUsesWithTargetAndMethod = function (target, methodName) {
        return this.useInterceptors.filter(function (use) {
            return use.target === target && use.method === methodName;
        });
    };
    /**
     * Filters parameters by a given classes.
     */
    MetadataArgsStorage.prototype.filterParamsWithTargetAndMethod = function (target, methodName) {
        return this.params.filter(function (param) {
            return param.object.constructor === target && param.method === methodName;
        });
    };
    /**
     * Filters response handlers by a given class.
     */
    MetadataArgsStorage.prototype.filterResponseHandlersWithTarget = function (target) {
        return this.responseHandlers.filter(function (property) {
            return property.target === target;
        });
    };
    /**
     * Filters response handlers by a given classes.
     */
    MetadataArgsStorage.prototype.filterResponseHandlersWithTargetAndMethod = function (target, methodName) {
        return this.responseHandlers.filter(function (property) {
            return property.target === target && property.method === methodName;
        });
    };
    /**
     * Removes all saved metadata.
     */
    MetadataArgsStorage.prototype.reset = function () {
        this.controllers = [];
        this.middlewares = [];
        this.interceptors = [];
        this.uses = [];
        this.useInterceptors = [];
        this.actions = [];
        this.params = [];
        this.responseHandlers = [];
    };
    return MetadataArgsStorage;
}());
exports.MetadataArgsStorage = MetadataArgsStorage;
